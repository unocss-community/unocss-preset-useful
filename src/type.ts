export interface UsefulOptions {
  /**
   * Extract rgba color in css variable
   *
   * @default false
   */
  unColor?: boolean | string

  /**
   * Expand theme animation name usage
   *
   * [ name, duration, timing-function, iteration-count ]
   *
   * @example
   * ```ts
    themeAnimate: ['spin 1s linear infinite'],
   * ```
   */
  themeAnimate?: string[]
}

export type ResolvedOptions = Required<UsefulOptions>

export type DeepPartial<T> = { [P in keyof T]?: DeepPartial<T[P]> }
