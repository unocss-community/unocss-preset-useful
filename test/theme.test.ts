import { createGenerator } from 'unocss'
import { describe, expect, it } from 'vitest'
import { presetUseful } from 'unocss-preset-useful'
import type { UsefulExtends, UsefulTheme } from '../packages/core/src/types'

describe('presetUseful theme', () => {
  const extend: UsefulExtends = {
    animation: {
      'spin': 'spin 1s linear infinite',
      'bounce': 'bounce 2s ease-in-out 3',
      'fade': 'fade 1s ease-in-out 3',
      'foo': 'foo 1s * 3',
      'bar': 'bar 1s +',
      'wiggle-slow': 'wiggle 1s ease-in-out infinite',
    },
    keyframes: {
      wiggle: {
        '0%, 100%': { transform: 'rotate(-3deg)' },
        '50%': { transform: 'rotate(3deg)' },
        '20%, 30%': { transform: 'rotate(-3deg)' },
        '80%': { transform: 'rotate(3deg)' },
      },
    },
  }

  it('theme extend configuration', async () => {
    const _uno = createGenerator<UsefulTheme>({
      presets: [
        presetUseful({
          enableDefaultShortcuts: false,
          enableMagicAnimations: false,
          uno: true,
          theme: {
            extend,
          },
        }),
      ],
    })
    const { css } = await _uno.generate('animate-wiggle-slow', { preflights: false })

    expect(_uno.config.theme.animation).toMatchSnapshot()
    expect(css).toMatchInlineSnapshot(`
      "/* layer: shortcuts */
      .animate-wiggle-slow{animation:wiggle 1s ease-in-out infinite;}
      /* layer: default */
      @keyframes wiggle{0%, 100%{transform:rotate(-3deg);}50%{transform:rotate(3deg);}20%, 30%{transform:rotate(-3deg);}80%{transform:rotate(3deg);}}"
    `)
  })
})
