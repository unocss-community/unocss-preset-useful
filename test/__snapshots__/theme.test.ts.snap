// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Theme > themeAnimate configuration 1`] = `
{
  "counts": {
    "bounce": "3",
    "fade": "3",
    "spin": "infinite",
  },
  "durations": {
    "bounce": "2s",
    "fade": "1s",
    "spin": "1s",
  },
  "timingFns": {
    "bounce": "ease-in-out",
    "fade": "ease-in-out",
    "spin": "linear",
  },
}
`;
